### PR TITLE
calling convention fixes for better consistency

### PR DESCRIPTION
**Pull request by DavidXanatos**
The changes helps to compile the phlib into projects set up for different default calling conventions (like Task Explorer)

### CODE REVIEW

File: kphlib/include/kphdyndata.h [hunk 1]
```diff
@@ -14,7 +14,7 @@
 #include <kphdyn.h>
 
 _Must_inspect_result_
-NTSTATUS KphDynDataGetConfiguration(
+NTSTATUS NTAPI KphDynDataGetConfiguration(
```
Comment (dmex) [inline kphlib/include/kphdyndata.h:17]: None of the Kph functions are exported or external.

File: phlib/include/apiimport.h [hunk 2]
```diff
@@ -229,7 +229,7 @@ typedef VOID (WINAPI* _UnsubscribeServiceChangeNotifications)(
     _In_ PSC_NOTIFICATION_REGISTRATION pSubscription
     );
 
-#define PH_DECLARE_IMPORT(Name) _##Name Name##_Import(VOID)
+#define PH_DECLARE_IMPORT(Name) _##Name NTAPI Name##_Import(VOID)
```
Comment (dmex) [inline phlib/include/apiimport.h:232]: These types don't have calling conventions.

File: phlib/include/appresolver.h [hunk 3]
```diff
@@ -14,23 +14,23 @@
 
 EXTERN_C_START
 
-HRESULT PhAppResolverGetAppIdForProcess(
+HRESULT NTAPI PhAppResolverGetAppIdForProcess(
```
Comment (dmex) [inline phlib/include/appresolver.h:17]: The AppResolver functions are not external or exported.

File: phlib/include/kphcomms.h [hunk 4]
```diff
@@ -14,7 +14,7 @@
 
 #include <kphmsg.h>
 
-NTSTATUS KphFilterLoadUnload(
+NTSTATUS NTAPI KphFilterLoadUnload(
```
Comment (dmex) [inline phlib/include/kphcomms.h:17]: These functions are not exported or external.

File: phlib/include/phutil.h [hunk 5]
```diff
@@ -1853,6 +1853,7 @@ PhHungWindowFromGhostWindow(
 
 PHLIBAPI
 NTSTATUS
+NTAPI
```
Comment (dmex) [inline phlib/include/phutil.h:1856]: ✔️

File: phlib/include/settings.h [hunk 6]
```diff
@@ -376,7 +377,9 @@ VOID PhResetSettings(
 // begin_phapppub
 // High-level settings creation
 
-VOID PhAddSetting(
+VOID
+NTAPI
+PhAddSetting(
```
Comment (dmex) [inline phlib/include/settings.h:382]: ✔️

**Review by dmex — CHANGES_REQUESTED** [review]

File: kphlib/include/kphdyndata.h (hunk 1 above)
Comment (jxy-s) [reply kphlib/include/kphdyndata.h:17]: Agreed, should not be necessary.

Comment (DavidXanatos) [reply kphlib/include/kphdyndata.h:17]: 3rd party projects which use the kphlib may link against un exported functions, and if they by default use a different calling convention then __stdcall a fix like this is required.

File: phlib/include/appresolver.h (hunk 3 above)
Comment (DavidXanatos) [reply phlib/include/appresolver.h:17]: 3rd party projects which use the kphlib may link against un exported functions, and if they by default use a different calling convention then __stdcall a fix like this is required.

File: phlib/include/kphcomms.h (hunk 4 above)
Comment (DavidXanatos) [reply phlib/include/kphcomms.h:17]: 3rd party projects which use the kphlib may link against un exported functions, and if they by default use a different calling convention then __stdcall a fix like this is required.

File: phlib/include/apiimport.h (hunk 2 above)
Comment (DavidXanatos) [reply phlib/include/apiimport.h:232]: well my VS2022 begs to differ without that fix i can not build a x86 version of Task Explorer

**Review by jxy-s — CHANGES_REQUESTED** [review]
Many of the calling convention changes aren't necessary upstream. The request is to support oddity calling conventions that might be default in a secondary project. These are internal libraries and not intended to be exported. If there is a use case of a secondary project, those should be maintained in an alternate fork that is specific to the oddities of that other project.

**Comment by DavidXanatos** [comment]
> Many of the calling convention changes aren't necessary upstream. The request is to support oddity calling conventions that might be default in a secondary project. These are internal libraries and not intended to be exported. If there is a use case of a secondary project, those should be maintained in an alternate fork that is specific to the oddities of that other project.

The calling convention use by VS by default is __cdecl so if anything the default use of __stdcall is an oddity.
Since SI is open source IMHO it is to be expected that other projects will use parts of it especially if they come prepackaged as convenient libraries.
For SI it is a change without a difference but for others its a great help hence I assumed it would be a good idea to be more explicit about the used calling convention.
Its disappointing that such a improvement is not welcome.

**Comment by jxy-s** [comment]
> > Many of the calling convention changes aren't necessary upstream. The request is to support oddity calling conventions that might be default in a secondary project. These are internal libraries and not intended to be exported. If there is a use case of a secondary project, those should be maintained in an alternate fork that is specific to the oddities of that other project.
> 
> The calling convention use by VS by default is __cdecl so if anything the default use of __stdcall is an oddity.
> Since SI is open source IMHO it is to be expected that other projects will use parts of it especially if they come prepackaged as convenient libraries.
> For SI it is a change without a difference but for others its a great help hence I assumed it would be a good idea to be more explicit about the used calling convention.
> Its disappointing that such a improvement is not welcome.

I'm not sure that's accurate. For kphlib, It's an x64 or ARM64 static library linked and compiled into the target binary. So the calling convention shouldn't be `__cdecl`. If it is, we should open an alternate discussion on correcting that problem separately from specifying it explicitly. The project might be misconfigured. But none of those changes would be in support of consuming that from a 3rd party.

**Comment by DavidXanatos** [comment]
> 
> I'm not sure that's accurate. For kphlib, it's an x64 or ARM64 static library linked and compiled into the target binary. So the calling convention shouldn't be `__cdecl`. If it is, we should open an alternate discussion on correcting that problem separately from specifying it explicitly. The project might be misconfigured. But none of those changes would be in support of consuming that from a 3rd party.

x64 and ARM64 each only supports one calling convention which booth somewhat resembles __vectorcall the explicit  calling convention specifier have only a meaning on x86, so the issue only arises when trying to compile a project using kphlib for x86, the others x64 and ARM64 work just fine no matter what.

Process Hacker opted to, for whatever reason, to use __stdcall in its projects, which as described is not the VS default nowadays new projects start out with __cdecl preset.
So any 3rd party project which want to use kphlib needs to workaround this mismatch.

If you don't want to support the use case of kphlib being linked by 3rd parties than this pull request is irrelevant and should be closed.

**Comment by jxy-s** [comment]
> > I'm not sure that's accurate. For kphlib, it's an x64 or ARM64 static library linked and compiled into the target binary. So the calling convention shouldn't be `__cdecl`. If it is, we should open an alternate discussion on correcting that problem separately from specifying it explicitly. The project might be misconfigured. But none of those changes would be in support of consuming that from a 3rd party.
> 
> x64 and ARM64 each only supports one calling convention which booth somewhat resembles __vectorcall the explicit calling convention specifier have only a meaning on x86, so the issue only arises when trying to compile a project using kphlib for x86, the others x64 and ARM64 work just fine no matter what.
> 
> Process Hacker opted to, for whatever reason, to use __stdcall in its projects, which as described is not the VS default nowadays new projects start out with __cdecl preset. So any 3rd party project which want to use kphlib needs to workaround this mismatch.
> 
> If you don't want to support the use case of kphlib being linked by 3rd parties than this pull request is irrelevant and should be closed.
> 
> PS: IMHO fostering 3rd party usage would help to engage more contributions to the project and be in the projects very own self interest.

The driver doesn't support 32-bit. We should probably stop referencing `kphlib` for that build and `ifdef` out all the client kph logic instead.

**Comment by DavidXanatos** [comment]
> The driver doesn't support 32-bit. We should probably stop referencing `kphlib` for that build and `ifdef` out all the client kph logic instead.

Only one of the changes addresses the kphlib all other changes address the user mode phlib.
If we could ensure phlib can be used by projects which don't use __stdcall as default that would already help a great lot :)

**Comment by jxy-s** [comment]
There has been no progress and the recent rewrite caused many conflicts. Closing.

**Pull request closed without merging**